hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/deployment/ResourceIterator.java

```diff
@@ -64,33 +64,41 @@ public void remove() {
 
     @Override
     public ResourcePart next() {
+        if (nextPart  == null) {
+            throw new NoSuchElementException();
+        }
         try {
-            if (inputStream == null) {
-                switchFile();
-            }
+            return nextPart;
+        } finally {
+            readNext();
+        }
+    }
 
-            byte[] bytes = read(inputStream, partSize);
+    private void readNext() {
+        try {
+            nextPart = null;
+            while (nextPart == null && (configIterator.hasNext() || inputStream != null)) {
+                if (inputStream == null) {
+                    switchFile();
+                }
 
-            if (inputStream.available() <= 0) {
-                close(inputStream);
-                inputStream = null;
-            }
-            if (bytes.length > 0) {
-                ResourcePart part = new ResourcePart(resourceConfig.getDescriptor(), bytes, offset);
-                offset += bytes.length;
-                return part;
-            } else {
-                throw new NoSuchElementException();
+                int bytesRead = inputStream.read(buffer);
+
+                if (bytesRead < 0) {
+                    bytesRead = 0;
+                    inputStream.close();
```

Comment:
It's not, it's also not, when we don't finish the iteration, if the exception is in our caller.